src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,472 @@
+            {
+                if (RunningOnCreatingThread && _appended.Append(element, _appCount))
+                {
+                    return new AppendPrependNIterator<TSource>(_source, _prepended, _preCount, _appended, _appCount + 1);
+                }
+
+                return new AppendPrepend1Iterator<TSource>(this, element, true);
+            }
+
+            public override AppendPrependIterator<TSource> Prepend(TSource element)
+            {
+                if (RunningOnCreatingThread)
+                {
+                    if (_prepended == null)
+                    {
+                        return new AppendPrependNIterator<TSource>(_source, new SharedCollection<TSource>(element), 1, _appended, _appCount);
+                    }
+
+                    if (_prepended.Prepend(element, _preCount))
+                    {
+                        return new AppendPrependNIterator<TSource>(_source, _prepended, _preCount + 1, _appended, _appCount);
+                    }
+                }
+
+                return new AppendPrepend1Iterator<TSource>(this, element, false);
+            }
+
+            public override int GetCount(bool onlyIfCheap)
+            {
+                IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                if (listProv != null)
+                {
+                    int count = listProv.GetCount(onlyIfCheap);
+                    return count == -1 ? -1 : count + _preCount + _appCount;
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + _preCount + _appCount : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count];
+                int index = 0;
+                for (int i = -_preCount; i != 0; ++i, ++index)
+                {
+                    array[index] = _prepended[i];
+                }
+
+                Debug.Assert(index == _preCount);
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == array.Length - _appCount);
+                for (int i = 0; i != _appCount; ++i, ++index)
+                {
+                    array[index] = _appended[i];
+                }
+
+                Debug.Assert(index == array.Length);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(Math.Max(_appCount + _preCount, 4)) : new List<TSource>(count);
```

Comment:
@stephentoub I dropped the `new List<TSource>(4)` from the other case, but I've left `Math.Max(_appCount + _preCount, 4)` as it seems that "the sum of the sizes of the two parts we know are adding in for sure, unless that's pretty low in which case a wee bit more" is a reasonable heuristic for the starting size, with 4 as a reasonable value for "wee bit more".